### ISSUE TITLE
lvalue sub shows "Bizarre copy of ARRAY in block exit" in very specific case.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13258]
Migrated from [rt.perl.org#119797](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=119797) (status was 'resolved')

Searchable as RT119797$

**Comment by p5pRT** [comment]
### From glitchmr@myopera.com

#### Created by glitchmr@myopera.com



Following code involving lvalue shows "Bizarre copy of ARRAY in block
exit"\. I've experimented with it\, and it seems to happen with any
array \(lexical or global\)\, as long the \`if\` condition involves array
somehow\, and returns false \(for example\, \`if \(@&#8203;array\, 0\)\`\)\, and \`else\`
block returns \`@&#8203;array\` without \`return\` keyword\.

&nbsp;    \# The subroutine has to be declared as lvalue\, so you can write to
&nbsp;    \# it in order to trigger the bug\.
&nbsp;    sub array : lvalue \{
&nbsp;        \# This test has to involve @&#8203;array somehow\, but fail\.
&nbsp;        if \(@&#8203;array\) \{
&nbsp;            \# The contents of if block don't matter\, so I left this
&nbsp;            \# block empty\. It just has to exist\.
&nbsp;        \}
&nbsp;        else \{
&nbsp;            \# The else condition has to return @&#8203;array\. Return keyword
&nbsp;            \# cannot be used for this bug to trigger\.
&nbsp;            @&#8203;array;
&nbsp;        \}
&nbsp;    \}
&nbsp;    \# Now\, to trigger the bug\, it's enough to write to array\.
&nbsp;    \(array\) = \(\);

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.18.1:

Configured by nobody at Tue Aug 13 13:22:58 CEST 2013.

Summary of my perl5 (revision 5 version 18 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=3.10.5-1-arch,
    archname=x86_64-linux-thread-multi
    uname='linux mnt-chroots-arch-extra-x86_64-flo-64 3.10.5-1-arch #1
    smp preempt mon aug 5 08:04:22 cest 2013 x86_64 gnulinux '
    config_args='-des -Dusethreads -Duseshrplib -Doptimize=-march=x86-64
    -mtune=generic -O2 -pipe -fstack-protector --param=ssp-buffer-size=4
    -Dprefix=/usr -Dvendorprefix=/usr
    -Dprivlib=/usr/share/perl5/core_perl
    -Darchlib=/usr/lib/perl5/core_perl
    -Dsitelib=/usr/share/perl5/site_perl
    -Dsitearch=/usr/lib/perl5/site_perl
    -Dvendorlib=/usr/share/perl5/vendor_perl
    -Dvendorarch=/usr/lib/perl5/vendor_perl
    -Dscriptdir=/usr/bin/core_perl -Dsitescript=/usr/bin/site_perl
    -Dvendorscript=/usr/bin/vendor_perl -Dinc_version_list=none
    -Dman1ext=1perl -Dman3ext=3perl -Dcccdlflags='-fPIC'
    -Dlddlflags=-shared -Wl,-O1,--sort-common,--as-needed,-z,relro
    -Dldflags=-Wl,-O1,--sort-common,--as-needed,-z,relro'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing
    -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE
    -D_FILE_OFFSET_BITS=64',
    optimize='-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector
    --param=ssp-buffer-size=4',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe
    -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.8.1 20130725 (prerelease)',
    gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
    lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-O1,--sort-common,--as-needed,-z,relro
    -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib
    /lib64 /usr/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.17.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.17'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E
    -Wl,-rpath,/usr/lib/perl5/core_perl/CORE'
    cccdlflags='-fPIC', lddlflags='-shared
    -Wl,-O1,--sort-common,--as-needed,-z,relro -L/usr/local/lib
    -fstack-protector'

Locally applied patches:
    


@INC for perl 5.18.1:
    /usr/lib/perl5/site_perl
    /usr/share/perl5/site_perl
    /usr/lib/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib/perl5/core_perl
    /usr/share/perl5/core_perl
    .


Environment for perl 5.18.1:
    HOME=/home/glitchmr
    LANG=pl_PL.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/fish

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Sep 15 03&#8203;:08&#8203;:16 2013\, glitchmr@&#8203;myopera\.com wrote&#8203;:
> 
> This is a bug report for perl from glitchmr@&#8203;myopera\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.18\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> Following code involving lvalue shows "Bizarre copy of ARRAY in block
> exit"\. I've experimented with it\, and it seems to happen with any
> array \(lexical or global\)\, as long the \`if\` condition involves array
> somehow\, and returns false \(for example\, \`if \(@&#8203;array\, 0\)\`\)\, and \`else\`
> block returns \`@&#8203;array\` without \`return\` keyword\.
> 
>     \# The subroutine has to be declared as lvalue\, so you can write to
>     \# it in order to trigger the bug\.
>     sub array : lvalue \{
>         \# This test has to involve @&#8203;array somehow\, but fail\.
>         if \(@&#8203;array\) \{
>             \# The contents of if block don't matter\, so I left this
>             \# block empty\. It just has to exist\.
>         \}
>         else \{
>             \# The else condition has to return @&#8203;array\. Return keyword
>             \# cannot be used for this bug to trigger\.
>             @&#8203;array;
>         \}
>     \}
>     \# Now\, to trigger the bug\, it's enough to write to array\.
>     \(array\) = \(\);
> 


1\. I cannot reproduce the error you cite\.  I cannot generate this error&#8203;:
 ""Bizarre copy of ARRAY in block exit"\.

2\. Did you try running your program with 'use strict; use warnings;'? 
If so\, you would have gotten messages that indicated problematic code\.

3\. Did you try rewriting your program to name the subroutine something
other than 'array'?  That only made for a more confusing bug report\.

Once I started to debug this with steps 2 and 3 above\, I found that I
could get an executable \(though not very useful\) program  by simply
placing '\(\)' after the subroutine call\.

\#\#\#\#\#\#\#\#\#\#
$ cat 119797\-lvalue\.pl 
my @&#8203;this = \(1\.\.5\);

\(that\(\)\) = \(\);
print "@&#8203;this\\n";

sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
\#\#\#\#\#\#\#\#\#\#
$ perl 119797\-lvalue\.pl 
1 2 3 4 5
\#\#\#\#\#\#\#\#\#\#

I don't think there is a bug here\.



> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=medium
> \-\-\-
> Site configuration information for perl 5\.18\.1&#8203;:
> 
> Configured by nobody at Tue Aug 13 13&#8203;:22&#8203;:58 CEST 2013\.
> 
> Summary of my perl5 \(revision 5 version 18 subversion 1\) configuration&#8203;:
>    
>   Platform&#8203;:
>     osname=linux\, osvers=3\.10\.5\-1\-arch\,
>     archname=x86\_64\-linux\-thread\-multi
>     uname='linux mnt\-chroots\-arch\-extra\-x86\_64\-flo\-64 3\.10\.5\-1\-arch \#1
>     smp preempt mon aug 5 08&#8203;:04&#8203;:22 cest 2013 x86\_64 gnulinux '
>     config\_args='\-des \-Dusethreads \-Duseshrplib \-Doptimize=\-march=x86\-64
>     \-mtune=generic \-O2 \-pipe \-fstack\-protector \-\-param=ssp\-buffer\-size=4
>     \-Dprefix=/usr \-Dvendorprefix=/usr
>     \-Dprivlib=/usr/share/perl5/core\_perl
>     \-Darchlib=/usr/lib/perl5/core\_perl
>     \-Dsitelib=/usr/share/perl5/site\_perl
>     \-Dsitearch=/usr/lib/perl5/site\_perl
>     \-Dvendorlib=/usr/share/perl5/vendor\_perl
>     \-Dvendorarch=/usr/lib/perl5/vendor\_perl
>     \-Dscriptdir=/usr/bin/core\_perl \-Dsitescript=/usr/bin/site\_perl
>     \-Dvendorscript=/usr/bin/vendor\_perl \-Dinc\_version\_list=none
>     \-Dman1ext=1perl \-Dman3ext=3perl \-Dcccdlflags='\-fPIC'
>     \-Dlddlflags=\-shared \-Wl\,\-O1\,\-\-sort\-common\,\-\-as\-needed\,\-z\,relro
>     \-Dldflags=\-Wl\,\-O1\,\-\-sort\-common\,\-\-as\-needed\,\-z\,relro'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     useithreads=define\, usemultiplicity=define
>     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>     use64bitint=define\, use64bitall=define\, uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing
>     \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
>     \-D\_FILE\_OFFSET\_BITS=64'\,
>     optimize='\-march=x86\-64 \-mtune=generic \-O2 \-pipe \-fstack\-protector
>     \-\-param=ssp\-buffer\-size=4'\,
>     cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-pipe
>     \-fstack\-protector \-I/usr/local/include'
>     ccversion=''\, gccversion='4\.8\.1 20130725 \(prerelease\)'\,
>     gccosandvers=''
>     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>     ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
>     lseeksize=8
>     alignbytes=8\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc'\, ldflags ='\-Wl\,\-O1\,\-\-sort\-common\,\-\-as\-needed\,\-z\,relro
>     \-fstack\-protector \-L/usr/local/lib'
>     libpth=/usr/local/lib /lib/\.\./lib /usr/lib/\.\./lib /lib /usr/lib
>     /lib64 /usr/lib64
>     libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
>     \-lgdbm\_compat
>     perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
>     libc=/lib/libc\-2\.17\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
>     gnulibc\_version='2\.17'
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E
>     \-Wl\,\-rpath\,/usr/lib/perl5/core\_perl/CORE'
>     cccdlflags='\-fPIC'\, lddlflags='\-shared
>     \-Wl\,\-O1\,\-\-sort\-common\,\-\-as\-needed\,\-z\,relro \-L/usr/local/lib
>     \-fstack\-protector'
> 
> Locally applied patches&#8203;:
>     
> 
> \-\-\-
> @&#8203;INC for perl 5\.18\.1&#8203;:
>     /usr/lib/perl5/site\_perl
>     /usr/share/perl5/site\_perl
>     /usr/lib/perl5/vendor\_perl
>     /usr/share/perl5/vendor\_perl
>     /usr/lib/perl5/core\_perl
>     /usr/share/perl5/core\_perl
>     \.
> 
> \-\-\-
> Environment for perl 5\.18\.1&#8203;:
>     HOME=/home/glitchmr
>     LANG=pl\_PL\.UTF\-8
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/usr/bin
>     PERL\_BADLANG \(unset\)
>     SHELL=/usr/bin/fish

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Sep 15 06&#8203;:21&#8203;:00 2013\, jkeenan wrote&#8203;:
> \#\#\#\#\#\#\#\#\#\#
> $ cat 119797\-lvalue\.pl 
> my @&#8203;this = \(1\.\.5\);
> 
> \(that\(\)\) = \(\);
> print "@&#8203;this\\n";
> 
> sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> \#\#\#\#\#\#\#\#\#\#
> $ perl 119797\-lvalue\.pl 
> 1 2 3 4 5
> \#\#\#\#\#\#\#\#\#\#
> 
> I don't think there is a bug here\.

Your version puts something in the array\, avoiding the else block\.

Try this&#8203;:

\#\!perl
my @&#8203;this = \(\);

\(that\(\)\) = \(\);
print "@&#8203;this\\n";

sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
\_\_END\_\_

That gives me a bizarre copy\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

On Sun\, Sep 15\, 2013 at 3&#8203;:21 PM\, James E Keenan via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> On Sun Sep 15 03&#8203;:08&#8203;:16 2013\, glitchmr@&#8203;myopera\.com wrote&#8203;:
>


> 1\. I cannot reproduce the error you cite\.  I cannot generate this error&#8203;:
>  ""Bizarre copy of ARRAY in block exit"\.
>

&nbsp;  I can\, with 5\.14\.2 and some rewriting of the code\.

&nbsp;  I don't have a more recent version readily available\, so it may well be
fixed now\, but \.\.\. essentially it's the same as the OP reported\, so here
goes&#8203;:


2\. Did you try running your program with 'use strict; use warnings;'?
> If so\, you would have gotten messages that indicated problematic code\.
>

&nbsp;  I did\.  They went away once I declared @&#8203;array and \.\.\. added some lvalue
\(even read\-only\) to the if\-branch \.\.\. perhaps 5\.14\.2 optimizes differently?


3\. Did you try rewriting your program to name the subroutine something
> other than 'array'?  That only made for a more confusing bug report\.
>
> Once I started to debug this with steps 2 and 3 above\, I found that I
> could get an executable \(though not very useful\) program  by simply
> placing '\(\)' after the subroutine call\.
>
> \#\#\#\#\#\#\#\#\#\#
> $ cat 119797\-lvalue\.pl
> my @&#8203;this = \(1\.\.5\);
>
> \(that\(\)\) = \(\);
> print "@&#8203;this\\n";
>
> sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> \#\#\#\#\#\#\#\#\#\#
> $ perl 119797\-lvalue\.pl
> 1 2 3 4 5
> \#\#\#\#\#\#\#\#\#\#
>
> I don't think there is a bug here\.
>

&nbsp;  Your rewrite gets two things wrong&#8203;:

&nbsp;  First\, to trigger the bug\, the else\-branch must be taken\, so @&#8203;this should
be empty or the conditional changed\.

&nbsp;  Second\, your need for the '\(\)' stems from declaring &that after using it
\.\.\. ;\-\)

&nbsp;  Additionally\, I added a \(read\-only\) lvalue to the if\-branch\, to avoid
perl 5\.14\.2 going down a different internal path \.\.\.

\#\#\#\#\#\#\#\#\#\#
use strict;
use warnings;

my @&#8203;this;

sub that :lvalue \{ if \(@&#8203;this\) \{ $& \} else \{ @&#8203;this; \} \}

\(that\) = \(\);
print "@&#8203;this\\n";
\#\#\#\#\#\#\#\#\#\#
perl /home/eirik/tmp/kladd\.pl
Bizarre copy of ARRAY in block exit at /home/eirik/tmp/kladd\.pl line 6\.
\#\#\#\#\#\#\#\#\#\#


Eirik

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Sep 15 07&#8203;:10&#8203;:48 2013\, sprout wrote&#8203;:
> On Sun Sep 15 06&#8203;:21&#8203;:00 2013\, jkeenan wrote&#8203;:
> > \#\#\#\#\#\#\#\#\#\#
> > $ cat 119797\-lvalue\.pl 
> > my @&#8203;this = \(1\.\.5\);
> > 
> > \(that\(\)\) = \(\);
> > print "@&#8203;this\\n";
> > 
> > sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> > \#\#\#\#\#\#\#\#\#\#
> > $ perl 119797\-lvalue\.pl 
> > 1 2 3 4 5
> > \#\#\#\#\#\#\#\#\#\#
> > 
> > I don't think there is a bug here\.
> 
> Your version puts something in the array\, avoiding the else block\.
> 
> Try this&#8203;:
> 
> \#\!perl
> my @&#8203;this = \(\);
> 
> \(that\(\)\) = \(\);
> print "@&#8203;this\\n";
> 
> sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> \_\_END\_\_
> 
> That gives me a bizarre copy\.
> 


Okay\.  So now we should ask&#8203;:  How significant is this bug?  What can we
do to fix it?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Sep 15 09&#8203;:10&#8203;:35 2013\, jkeenan wrote&#8203;:
> On Sun Sep 15 07&#8203;:10&#8203;:48 2013\, sprout wrote&#8203;:
> > On Sun Sep 15 06&#8203;:21&#8203;:00 2013\, jkeenan wrote&#8203;:
> > > \#\#\#\#\#\#\#\#\#\#
> > > $ cat 119797\-lvalue\.pl 
> > > my @&#8203;this = \(1\.\.5\);
> > > 
> > > \(that\(\)\) = \(\);
> > > print "@&#8203;this\\n";
> > > 
> > > sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> > > \#\#\#\#\#\#\#\#\#\#
> > > $ perl 119797\-lvalue\.pl 
> > > 1 2 3 4 5
> > > \#\#\#\#\#\#\#\#\#\#
> > > 
> > > I don't think there is a bug here\.
> > 
> > Your version puts something in the array\, avoiding the else block\.
> > 
> > Try this&#8203;:
> > 
> > \#\!perl
> > my @&#8203;this = \(\);
> > 
> > \(that\(\)\) = \(\);
> > print "@&#8203;this\\n";
> > 
> > sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> > \_\_END\_\_
> > 
> > That gives me a bizarre copy\.
> > 
> 
> 
> Okay\.  So now we should ask&#8203;:  How significant is this bug?

It’s low priority\, as the workaround \(explicit return\) is simple\.

> What can we
> do to fix it?

I suspect skipping the leave op when followed by leavesublv would work\.
 In fact\, skipping it for any sub \(not just lvalue\) would probably make
implicit return faster in general\.

That would entail nulling it \(via op\_null\(\.\.\.\)\) in S\_finalize\_op in op\.c\.

\(That’s without looking at the code\, so it could be all wrong\.\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From glitchmr@myopera.com

On Sun\, Sep 15\, 2013\, at 03&#8203;:21 PM\, James E Keenan via RT wrote&#8203;:
> 1\. I cannot reproduce the error you cite\.  I cannot generate this error&#8203;:
>  ""Bizarre copy of ARRAY in block exit"\.
> 
> 2\. Did you try running your program with 'use strict; use warnings;'? 
> If so\, you would have gotten messages that indicated problematic code\.
> 
> 3\. Did you try rewriting your program to name the subroutine something
> other than 'array'?  That only made for a more confusing bug report\.

Yes\, I used "use strict" and "use warnings"\, but it changes nothing\. I
can reproduce it using perlbot on Freenode \(Perl 5\.018001\)\.

\<GlitchMr> eval&#8203;: sub array : lvalue \{ if \(@&#8203;array\) \{\} else \{ @&#8203;array
\} \} \(array\) = \(\)
\<perlbot> GlitchMr&#8203;: ERROR&#8203;: Bizarre copy of ARRAY in block exit at \(eval
1833\) line 1\.

I can also reproduce it on Codepad \(using old version\, 5\.008\, but it
confirms that this bug existed here for here for a long time\) \-
\<http&#8203;://codepad\.org/x0KwLtUJ> \(it says leave\, but I blame old Perl
version\)\, and ideone \(with Perl 5\.016002\) \- \<http&#8203;://ideone\.com/02TLL7>\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Sep 15 11&#8203;:26&#8203;:14 2013\, sprout wrote&#8203;:
> On Sun Sep 15 09&#8203;:10&#8203;:35 2013\, jkeenan wrote&#8203;:
> > On Sun Sep 15 07&#8203;:10&#8203;:48 2013\, sprout wrote&#8203;:
> > > \#\!perl
> > > my @&#8203;this = \(\);
> > > 
> > > \(that\(\)\) = \(\);
> > > print "@&#8203;this\\n";
> > > 
> > > sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> > > \_\_END\_\_
> > > 
> > > That gives me a bizarre copy\.

The bug is worse than that\.  The cases that do not get a bizarre copy do
not return assignable arrays\, either&#8203;:

use Test&#8203;::More 'no\_plan';
sub if119797 : lvalue \{
&nbsp;    if \($\_\[0\]\) \{
&nbsp;    @&#8203;119797
&nbsp;    \}
\}
eval \{ \(if119797\(1\)\) = 4\.\.6 \};
is $@&#8203;\, ""\, '$@&#8203; after writing to array returned by if';
is "@&#8203;119797"\, "4 5 6"\, 'writing to array returned by if';

Output&#8203;:

ok 1 \- $@&#8203; after writing to array returned by if
not ok 2 \- writing to array returned by if
\#   Failed test 'writing to array returned by if'
\#   at \- line 9\.
\#          got&#8203;: ''
\#     expected&#8203;: '4 5 6'
1\.\.2
\# Looks like you failed 1 test of 2\.

> > 
> > 
> > Okay\.  So now we should ask&#8203;:  How significant is this bug?
> 
> It’s low priority\, as the workaround \(explicit return\) is simple\.
> 
> > What can we
> > do to fix it?
> 
> I suspect skipping the leave op when followed by leavesublv would work\.
>  In fact\, skipping it for any sub \(not just lvalue\) would probably make
> implicit return faster in general\.
> 
> That would entail nulling it \(via op\_null\(\.\.\.\)\) in S\_finalize\_op in op\.c\.
> 
> \(That’s without looking at the code\, so it could be all wrong\.\)

It is wrong\.  leavesub and leavesublv don’t pop scopes back to the
subroutine the way return does\, so the op\_leave preceding it is not
optional\.

The problem here has also to do with op trees not getting flagged properly&#8203;:

$ \./perl \-Ilib \-MO=Concise\,if119797 \-e 'sub if119797 : lvalue \{ if
\($\_\[0\]\) \{ @&#8203;119797 \} \}'
main&#8203;::if119797&#8203;:
6  \<1> leavesublv\[1 ref\] K/REFC\,1 \->\(end\)
\-     \<@&#8203;> lineseq KP \->6
1        \<;> nextstate\(main 3 \-e&#8203;:1\) v \->2
\-        \<1> null K/1 \->\-
3           \<|> and\(other\->4\) K/1 \->6
\-              \<1> ex\-aelem sK/2 \->3
\-                 \<1> ex\-rv2av sKR/1 \->\-
2                    \<$> aelemfast\(\*\_\) s \->3
\-                 \<0> ex\-const s \->\-
\-              \<@&#8203;> scope K \->\-
\-                 \<0> ex\-nextstate v \->4
5                 \<1> rv2av\[t1\] K/1 \->6
4                    \<$> gv\(\*119797\) s \->5
\-e syntax OK

Notice that the rv2av has no M indicating modifiable context\.

And then S\_adjust\_stack\_on\_leave in pp\_ctl\.c probably also needs adjustment\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Oct 23 06&#8203;:08&#8203;:23 2013\, sprout wrote&#8203;:
> Notice that the rv2av has no M indicating modifiable context\.

I misspoke\.  What it normally gets is LVSUB&#8203;:

$ \./perl \-Ilib \-MO=Concise\,if119797 \-e 'sub if119797 : lvalue \{  @&#8203;119797 \}'
main&#8203;::if119797&#8203;:
4  \<1> leavesublv\[1 ref\] K/REFC\,1 \->\(end\)
\-     \<@&#8203;> lineseq KP \->4
1        \<;> nextstate\(main 1 \-e&#8203;:1\) v \->2
3        \<1> rv2av\[t1\] K/LVSUB\,1 \->4
2           \<$> gv\(\*119797\) s \->3

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Oct 23 06&#8203;:08&#8203;:23 2013\, sprout wrote&#8203;:
> On Sun Sep 15 11&#8203;:26&#8203;:14 2013\, sprout wrote&#8203;:
> > On Sun Sep 15 09&#8203;:10&#8203;:35 2013\, jkeenan wrote&#8203;:
> > > On Sun Sep 15 07&#8203;:10&#8203;:48 2013\, sprout wrote&#8203;:
> > > > \#\!perl
> > > > my @&#8203;this = \(\);
> > > > 
> > > > \(that\(\)\) = \(\);
> > > > print "@&#8203;this\\n";
> > > > 
> > > > sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> > > > \_\_END\_\_
> > > > 
> > > > That gives me a bizarre copy\.
> 
> The bug is worse than that\.  The cases that do not get a bizarre copy do
> not return assignable arrays\, either&#8203;:
> 
> use Test&#8203;::More 'no\_plan';
> sub if119797 : lvalue \{
>     if \($\_\[0\]\) \{
>     @&#8203;119797
>     \}
> \}
> eval \{ \(if119797\(1\)\) = 4\.\.6 \};
> is $@&#8203;\, ""\, '$@&#8203; after writing to array returned by if';
> is "@&#8203;119797"\, "4 5 6"\, 'writing to array returned by if';

I have fixed both those issues in commit 2ec7f6f242\.

But it turns out that bare blocks exhibit similar problems\.  I will
probably fix that tomorrow\.

Also\, should while and for loops be able to return lvalues?  I have
always been confused as to whether the last statement inside a loop can
be implicitly returned\, so I have simply avoided depending on that\.  I
don’t know the answer here\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From glitchmr@myopera.com

Dnia środa\, 23 października 2013 21&#8203;:34&#8203;:26 piszesz&#8203;:
> Also\, should while and for loops be able to return lvalues?  I have
> always been confused as to whether the last statement inside a loop can
> be implicitly returned\, so I have simply avoided depending on that\.  I
> don’t know the answer here\.

I checked\. For some reason\, subroutine or \`do\` block ending with \`while\` block 
or \`for\` block always returns empty string \(not \`undef\` or last statement\, as 
I could've expected\)\. This is probably strange unintentional value\, but I 
think it may be kept here for backwards compatibility\.

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Konrad Borowski wrote&#8203;:
>I checked\. For some reason\, subroutine or \`do\` block ending with \`while\` block 
>or \`for\` block always returns empty string

The standard false value is an empty string\.  The last thing evaluated
by a while statement is the condition that comes up false to cause
termination of the loop\.  The return value is normally the actual
false value used&#8203;:

$ perl \-lwe 'print do \{ while\($\_\) \{\} \} // "undef" for 0\, undef'
0
undef

but it wasn't always thus when constant folding got involved&#8203;:

$ perl5\.16\.3 \-lwe 'print do \{ while\(0\) \{\} \} // "undef"' 
undef
$ perl5\.18\.1 \-lwe 'print do \{ while\(0\) \{\} \} // "undef"' 
0

\-zefram

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Oct 23 21&#8203;:34&#8203;:26 2013\, sprout wrote&#8203;:
> On Wed Oct 23 06&#8203;:08&#8203;:23 2013\, sprout wrote&#8203;:
> > On Sun Sep 15 11&#8203;:26&#8203;:14 2013\, sprout wrote&#8203;:
> > > On Sun Sep 15 09&#8203;:10&#8203;:35 2013\, jkeenan wrote&#8203;:
> > > > On Sun Sep 15 07&#8203;:10&#8203;:48 2013\, sprout wrote&#8203;:
> > > > > \#\!perl
> > > > > my @&#8203;this = \(\);
> > > > > 
> > > > > \(that\(\)\) = \(\);
> > > > > print "@&#8203;this\\n";
> > > > > 
> > > > > sub that :lvalue \{ if \(@&#8203;this\) \{ \} else \{ @&#8203;this; \} \}
> > > > > \_\_END\_\_
> > > > > 
> > > > > That gives me a bizarre copy\.
> > 
> > The bug is worse than that\.  The cases that do not get a bizarre copy do
> > not return assignable arrays\, either&#8203;:
> > 
> > use Test&#8203;::More 'no\_plan';
> > sub if119797 : lvalue \{
> >     if \($\_\[0\]\) \{
> >     @&#8203;119797
> >     \}
> > \}
> > eval \{ \(if119797\(1\)\) = 4\.\.6 \};
> > is $@&#8203;\, ""\, '$@&#8203; after writing to array returned by if';
> > is "@&#8203;119797"\, "4 5 6"\, 'writing to array returned by if';
> 
> I have fixed both those issues in commit 2ec7f6f242\.
> 
> But it turns out that bare blocks exhibit similar problems\.  I will
> probably fix that tomorrow\.

Now done with commit a373464fd\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
